lib/chef/resource/windows_user_privilege.rb

```diff
@@ -0,0 +1,159 @@
+        name_property: true
+
+      property :users, Array,
+        description: "An optional property to set the privilege for given users. Use only with set action."
+
+      property :privilege, [Array, String],
+        description: "Privilege to set for users.",
+        required: true,
+        callbacks: {
+           "Option privilege must include any of the: #{privilege_opts}" => lambda {
+             |v| v.is_a?(Array) ? (privilege_opts & v).size == v.size : privilege_opts.include?(v)
+           },
+         }
+
+      property :sensitive, [TrueClass, FalseClass], default: true
+
+      load_current_value do |new_resource|
+        unless new_resource.principal.nil?
+          privilege Chef::ReservedNames::Win32::Security.get_account_right(new_resource.principal) unless new_resource.action.include?(:set)
+        end
+      end
+
+      action :add do
+        ([*new_resource.privilege] - [*current_resource.privilege]).each do |user_right|
+          converge_by("adding user privilege #{user_right}") do
+            Chef::ReservedNames::Win32::Security.add_account_right(new_resource.principal, user_right)
+          end
+        end
+      end
+
+      action :set do
+        if new_resource.users.nil? || new_resource.users.empty?
+          raise Chef::Exceptions::ValidationFailed, "Users are required property with set action."
+        end
+
+        users = []
+
+        # Getting users with its domain for comparison
+        new_resource.users.each do |user|
+          user = Chef::ReservedNames::Win32::Security.lookup_account_name(user)
+          users << user[1].account if user
+        end
+
+        new_resource.privilege.each do |privilege|
+          accounts = Chef::ReservedNames::Win32::Security.get_account_with_user_rights(privilege)
+
+          # comparing the existing accounts for privilege with users
+          unless users == accounts
+            accounts.each do |account|
+              converge_by("removing user #{account[1]} from privilege #{privilege}") do
+                Chef::ReservedNames::Win32::Security.remove_account_right(account[1], privilege)
+              end
+            end
+
+            new_resource.users.each do |user|
+              converge_by("adding user #{user} to privilege #{privilege}") do
+                Chef::ReservedNames::Win32::Security.add_account_right(user, privilege)
+              end
+            end
+          end
+        end
+      end
+
+      action :remove do
+        curr_res_privilege = current_resource.privilege
+        new_res_privilege = new_resource.privilege
+
+        new_res_privilege = [] << new_res_privilege if new_resource.privilege.is_a?(String)
+        missing_res_privileges = (new_res_privilege - curr_res_privilege)
+
+        unless missing_res_privileges.empty?
+          Chef::Log.info("Privilege: #{missing_res_privileges.join(", ")} not present. Unable to delete")
+        end
```

Comment:
should that throw and fail?

Reply:
Not really. I will update the message which is little misleading. Remove action removes the given user from the privilege. So this just informs the user that we not found record/entry for the given combination so not removing it for e.g. "SeCreateSymbolicLinkPrivilege" if does not have "Administraor" user assigned already and we are trying to remove the same then it will just give the information to user and proceed.

Reply:
```
Recipe: windows_user_privilege::default
  * windows_user_privilege[Administrator] action remove[2020-02-24T22:42:45-08:00] INFO: Processing windows_user_privil
ge[Administrator] action remove (windows_user_privilege::default line 29)
[2020-02-24T22:42:45-08:00] INFO: User 'Administrator' for Privilege: SeCreatePagefilePrivilege, SeCreateSymbolicLinkPr
vilege not found. Nothing to remove.
 (up to date)
[2020-02-24T22:42:45-08:00] WARN: Skipping final node save because override_runlist was given
[2020-02-24T22:42:45-08:00] INFO: Chef Infra Client Run complete in 0.16102 seconds
[2020-02-24T22:42:45-08:00] INFO: Skipping removal of unused files from the cache

Running handlers:
[2020-02-24T22:42:45-08:00] INFO: Running report handlers
Running handlers complete
[2020-02-24T22:42:45-08:00] INFO: Report handlers complete
Chef Infra Client finished, 0/1 resources updated in 12 seconds
```

```
Recipe: windows_user_privilege::default
  * windows_user_privilege[Administrator] action add[2020-02-24T22:45:42-08:00] INFO: Processing windows_user_privilege[
Administrator] action add (windows_user_privilege::default line 13)

    - adding user privilege SeCreatePagefilePrivilege
  * windows_user_privilege[Administrator] action remove[2020-02-24T22:45:42-08:00] INFO: Processing windows_user_privile
ge[Administrator] action remove (windows_user_privilege::default line 29)
[2020-02-24T22:45:42-08:00] INFO: User 'Administrator' for Privilege: SeCreateSymbolicLinkPrivilege not found. Nothing t
o remove.

    - removing user privilege SeCreatePagefilePrivilege
[2020-02-24T22:45:42-08:00] WARN: Skipping final node save because override_runlist was given
[2020-02-24T22:45:42-08:00] INFO: Chef Infra Client Run complete in 0.1719867 seconds
[2020-02-24T22:45:42-08:00] INFO: Skipping removal of unused files from the cache

Running handlers:
[2020-02-24T22:45:42-08:00] INFO: Running report handlers
Running handlers complete
[2020-02-24T22:45:42-08:00] INFO: Report handlers complete
Chef Infra Client finished, 2/2 resources updated in 12 seconds
```